File: ApemanPlugin/src/main/kotlin/apeman_core/utils/BlocksUtils.java
package apeman_core.utils;

import apeman_core.base_entities.BlockOfMethod;
import com.intellij.psi.*;

import java.lang.reflect.Array;
import java.util.HashSet;
import java.util.Set;

public class BlocksUtils {
    private BlocksUtils() {}

    public static <T> Set<T> getElementsOfBlock(
            BlockOfMethod block, Class<T> aClassElement) {
        Set<T> result = new HashSet<>();

        for (int i = 0; i < block.getStatementsCount(); i++) {
            block.get(i).accept(
                    new JavaRecursiveElementVisitor() {
                        @Override
                        public void visitElement(PsiElement element) {
                            super.visitElement(element);
                            if (aClassElement.isAssignableFrom(element.getClass())) {
                                result.add(aClassElement.cast(element));
                            }
                        }
                    }
            );
        }
        return result;
    }

    private static int ourCount = 0;
    public static <T> int getCountOfElementFromBlock(BlockOfMethod block, T ourElement) {
        ourCount = 0;

        for (int i = 0; i < block.getStatementsCount(); i++) {
            block.get(i).accept(new JavaRecursiveElementVisitor() {
                @Override
                public void visitReferenceElement(PsiJavaCodeReferenceElement reference) {
                    super.visitReferenceElement(reference);

                    if (reference.isReferenceTo((PsiElement) ourElement)) {
                        ourCount++;
                    }
                }

                @Override
                public void visitElement(PsiElement element) {
                    super.visitElement(element);
                    if (ourElement == element) {
                        ourCount++;
                    }
                }
            });
        }
        return ourCount;
    }

    public static <T> double getFreqOfElementFromBlock(BlockOfMethod block, T element) {
        int count = getCountOfElementFromBlock(block, element);
        return (double)count / getNumStatementsRecursively(block);
    }

    private static int ourStatementsCount = 0;
    public static int getNumStatementsRecursively(BlockOfMethod block) {
        ourStatementsCount = 0;
        for (int i = 0; i < block.getStatementsCount(); i++) {
            block.get(i).accept(new JavaRecursiveElementVisitor() {


                @Override
                public void visitStatement(PsiStatement statement) {
                    super.visitStatement(statement);
                    if (statement.getParent() instanceof PsiCodeBlock) {
                        ourStatementsCount++;
                    }
                }
            });
        }
        return ourStatementsCount;
    }

    public static BlockOfMethod getBlockFromMethod(PsiMethod method) {
        PsiStatement[] statements = method.getBody().getStatements();
        return new BlockOfMethod(statements);
    }
}

File: ApemanPlugin/src/main/kotlin/apeman_core/utils/BlocksUtils.kt
package apeman_core.utils

import apeman_core.base_entities.BlockOfMethod
import com.intellij.psi.*

import java.lang.reflect.Array
import java.util.HashSet

object BlocksUtils {

    private var ourCount = 0

    private var ourStatementsCount = 0

    fun <T> getElementsOfBlock(block: BlockOfMethod, aClassElement: Class<T>): Set<T> {
        val result = HashSet<T>()

        for (i in 0 until block.statementsCount) {
            block[i].accept(
                    object : JavaRecursiveElementVisitor() {
                        override fun visitElement(element: PsiElement?) {
                            super.visitElement(element)
                            if (aClassElement.isAssignableFrom(element!!.javaClass)) {
                                result.add(aClassElement.cast(element))

                            } else if (element is PsiReference) {
                                val resolved = element.resolve() ?: return
                                if (aClassElement.isAssignableFrom(resolved.javaClass)) {
                                    result.add(aClassElement.cast(resolved))
                                }
                            }
                        }
                    }
            )
        }
        return result
    }

    fun <T> getCountOfElementFromBlock(block: BlockOfMethod, ourElement: T): Int {
        ourCount = 0

        for (i in 0 until block.statementsCount) {
            block[i].accept(object : JavaRecursiveElementVisitor() {
                override fun visitReferenceElement(reference: PsiJavaCodeReferenceElement) {
                    super.visitReferenceElement(reference)

                    if (reference.isReferenceTo(ourElement as PsiElement)) {
                        ourCount++
                    }
                }

                override fun visitElement(element: PsiElement?) {
                    super.visitElement(element)
                    if (ourElement === element) {
                        ourCount++
                    }
                }
            })
        }
        return ourCount
    }

    fun getNumStatementsRecursively(block: BlockOfMethod): Int {
        ourStatementsCount = 0
        for (i in 0 until block.statementsCount) {
            block[i].accept(object : JavaRecursiveElementVisitor() {

                override fun visitStatement(statement: PsiStatement) {
                    super.visitStatement(statement)
                    if (statement.parent is PsiCodeBlock) {
                        ourStatementsCount++
                    }
                }
            })
        }
        return ourStatementsCount
    }

    fun getStatementsRecursively(block: BlockOfMethod): Set<PsiStatement> {
        val result = hashSetOf<PsiStatement>()

        for (i in 0 until block.statementsCount) {
            block[i].accept(object : JavaRecursiveElementVisitor() {

                override fun visitStatement(statement: PsiStatement) {
                    super.visitStatement(statement)
                    if (statement.parent is PsiCodeBlock) {
                        result.add(statement)
                    }
                }
            })
        }
        return result
    }

    fun <T> getStatementsWithElement(block: BlockOfMethod, element: T): Int {
        val statementsMap = hashSetOf<PsiStatement>()
        var currentStatement: PsiStatement? = null

        for (i in 0 until block.statementsCount) {
            block[i].accept(object : JavaRecursiveElementVisitor() {

                override fun visitStatement(statement: PsiStatement?) {
                    val oldStatement = currentStatement
                    currentStatement = statement
                    super.visitStatement(statement)
                    currentStatement = oldStatement
                }

                override fun visitReferenceElement(reference: PsiJavaCodeReferenceElement) {
                    super.visitReferenceElement(reference)

                    if (reference.isReferenceTo(element as PsiElement)) {
                        statementsMap.add(currentStatement!!)
                    }
                }
            })
        }
        return statementsMap.count()
    }

    fun getBlockFromMethod(method: PsiMethod): BlockOfMethod {
        val statements = method.body!!.statements
        return BlockOfMethod(statements)
    }
}
